### ISSUE TITLE
weekly useful materials -05/11-

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#51]

**Comment by GENZITSU** [comment]
## 「AIの収益化」を成功させる方法

「MIT Sloan Management Review」とBCGが共同で調査した内容のシェア

>AIで財務上大きな価値を生み出している企業は10社に1社しかないと指摘している。

>財務上の価値を生み出すには次のような秘訣（ひけつ）があると考えられる。
・機械が自律的に学習するだけでなく、人間が絶えず機械に教え、人間も機械から絶えず教えを受ける
・コンテキストに応じて人間と機械が相互に関わり合う方法を幾つも考案する
・AIを利用した結果として、組織全体で学んだ内容に応じて幅広くプロセスを変える

human-in-the-loopを意識する必要がある。AIと実オペレーションの結合がまだ足りていない。  
実オペレーションに組み込めるだけの精度が足りてないとも言える...?

>　「AIには多額の投資が必要だ。サイロの中で取り組むだけでは規模の経済は得られない。相乗効果を生かすこともできず、コスト上のメリットも生じない。つまり、桁違いのコストがかかるビジネスモデルになることが多い」（セマク氏）

基本的にAIはコストが高いソリューションであることを意識する必要がある。  
横展開などができないとコストメリットを享受できない。。

>　もう一つ重要な問題がある。それは、大半の企業がAIの使い方を「誤って」いることだ。新しい収益の流れを生み出すことを目的とせずに、内部プロセスや運用手順の効率向上に集中している。

私位収益の流れを生み出すのってかなり難しそうだが...?  
1 オペレーションの改善に留まらないようにするってのが大事なのかな..?

> AIを導入する企業が抱える主な課題は、データ、そのデータのクリーンさと正確さ、そしてそのデータがどの程度「目的に沿っている」かだ。この場合の重要な問題は、適切なデータモデルを開発してトレーニングするには時間と労力を要することだ。

目的/課題に応じたデータを用意するところが一番難しいのかもしれない。  
多くのテックジャイアンとがデータ収集方法やユーザーからのFBをアプリケーションデザインに組み入れるのは、ここの課題感ゆえか...?  
PoCしておしまいではなく、データの適切なライフサイクルというかパイプラインをデザインすることも必要かも。

> 　「現時点での大きな障害は、AIを運用可能にし、運用環境に投入して、その環境内でAIの妥当性を維持する方法だ」

AIの運用を簡易化し、実環境からのFBを得ることがかなり重要。  
簡易なモデルでも良いので実環境に投入し、得られたFB(データ)をもとにモデル/ソリューションを改善する方が近道かも

> 　「だが、データモデルのトレーニング方法、運用中にモデルを変更する方法、デプロイ後にそのライフサイクルを管理する方法を把握することも必要だ。しかも、それは一度限りのことではない。データは絶えず変化する。」（イーガー氏）

モデルのデプロイは始まりにすぎない。
管理のためにはモデルとデータを一緒に管理する必要があるのかな。


[出典リンク](https://news.yahoo.co.jp/articles/7a265231131cdc3f0434a48b7b584a6128c0b215)

**Comment by GENZITSU** [comment]
## 学習済みTransformerモデルを使ったData Augmentation

元のテキストデータに[MASK]を付与して、学習済みTransformerの出力を用いてテキストのDAを行う。

言うはやすしだけど、DAに時間がかかりすぎるのがつらそう...。  
こいつが有効なのは、学習にはいくらでも時間かけていいけど、推論は高速にしたいって時かな？

実装例はこちら
```
class TransformerAugmenter():
    def __init__(self, model_name, num_sample_tokens, num_replace_tokens):
        self.model_name = model_name
        self.num_sample_tokens = num_sample_tokens
        self.num_replace_tokens = num_replace_tokens
        self.fill_mask = pipeline(
                    "fill-mask",
                    topk=self.num_sample_tokens,
                    model=self.model_name
                )
        self.mask_token = self.fill_mask.tokenizer.mask_token

    def _generate(self, text):
        text_replaced = text
        for _ in range(self.num_replace_tokens):
            replace_w = random.choice(text_replaced.split())
            text_replaced = text_replaced.replace(replace_w, self.mask_token, 1)

        aug_texts = []
        for aug_text in self.fill_mask(text_replaced):
            _text = aug_text["sequence"].replace("[CLS] ", "").replace(" [SEP]", "")
            if _text == text:
                continue
            aug_texts.append(_text)
        
        return aug_texts

    def generate(self, texts, targets):
        aug_texts = copy.deepcopy(texts)
        aug_targets = copy.deepcopy(targets)
        for text, target in tqdm(zip(texts, targets), total=len(texts)):
            _text = self._generate(text)
            aug_texts += _text
            aug_targets += [target for _ in range(len(_text))]
        return aug_texts, aug_targets


transformer_augmenter = TransformerAugmenter(
    model_name="cl-tohoku/bert-base-japanese-whole-word-masking",
    num_sample_tokens=5,  # 生成する文書数
    num_replace_tokens=1,  # 置換する単語数
)
```

TF-IDF + logistic regressionというシンプルなモデルで精度向上を検証。  
データ数1000件+DAがデータ数3000件に匹敵するとのこと。

<img width="300" alt="5c7e43bd8e72ad4c12b209b45e20afa0" src="https://user-images.githubusercontent.com/19774756/117988674-9ad38b80-b376-11eb-8e8a-6f0318e6a917.png"> <img width="300" alt="5c7e43bd8e72ad4c12b209b45e20afa0" src="https://user-images.githubusercontent.com/19774756/117988692-9dce7c00-b376-11eb-9d29-51606a4264c3.png">

[出典リンク](https://www.ai-shift.co.jp/techblog/1939)

**Comment by GENZITSU** [comment]
## 合成データ生成による多言語QAモデルの精度向上

背景: QAデータセットは英語に偏っているため、多言語モデルの精度が低い。とはいえ多言語のアノテーションは流石にだるい。

手法: 英語QAデータセット + wikiの記事500 を 機械翻訳モデルで他言語に翻訳して、question generationという技術を使って、擬似的な多言語QAデータセットを作る。

肝心のquestion generationは、context paragraphとanswerを所与のものとして、相応しい質問を生成すると言うもの。やってることはseq-2seqで、pretrained LMのおかげ近年精度が上がっているらしい[これとか](https://arxiv.org/pdf/1905.03197v3.pdf)。

どうもsquad とかsquad+機械翻訳で学習したquestion generation モデルを使って、wikiからqaを作っているらしいが、そんなことできるのか...?
いまいち、やり方がわからない。。。

**追記
[この論文](https://www.aclweb.org/anthology/P19-1620.pdfs)を見てる、一つのQAデータセットをもとに、
1. context→answer生成器
2. context+answer → question生成器
3. context+question → answer生成器
をつくることで、contextオンリーで擬似QAデータセットを作れるらしい。


結果: 何もしないよりかははるかに良くなる。

![スクリーンショット 2021-05-15 0 25 38](https://user-images.githubusercontent.com/19774756/118293098-436b2200-b514-11eb-9f9c-854f0c68ec2c.png)


![スクリーンショット 2021-05-15 0 26 07](https://user-images.githubusercontent.com/19774756/118293081-4108c800-b514-11eb-9c2b-acc907af386b.png)



[出典リンク](https://arxiv.org/abs/2010.12643v1)

**Comment by GENZITSU** [comment]
## efficientnet-v2のコードリリース

以下二つのフレームワークでのリリースが開始
- [tensorflow版](https://github.com/google/automl/tree/master/efficientnetv2)
- [pytorch/timm版](https://github.com/rwightman/pytorch-image-models/releases/tag/v0.1-effv2-weights)

てかefficientnet-v2ってなによ？

efficientnetの3つの問題点を解消したモデル
- 入力画像サイズが大きいと学習が遅くなる
- モデルの先頭付近にDepthwise畳み込みがあると学習が遅い
- モデルを均等にスケールアップするのは最適ではない

### 入力画像サイズが大きいと学習が遅くなる問題
背景: 入力画像サイズが大きいと、小さいバッチサイズでメモリがいっぱいになってしまうため、学習時間がかかる

そのため、入力画像サイズを徐々に小さいものから大きいものにかえるprogressive learningを導入。  
progressive learning自体は新しい手法ではないが、精度が出ていなかったとのこと。  
原因は小さい画像に対しても強いaugmentationをかけていたからとして、「入力画像サイズ」「augmentationの強さ」を徐々に大きくする手法を提案した。  
この学習手法は別のモデルにも有効なことがわかっている

<img width="400" alt="スクリーンショット 2021-05-15 21 23 18" src="https://user-images.githubusercontent.com/19774756/118360834-dd3dd800-b5c3-11eb-9366-b9f6b3304af0.png">

<img width="400" alt="スクリーンショット 2021-05-15 21 45 33" src="https://user-images.githubusercontent.com/19774756/118361544-f005dc00-b5c6-11eb-8245-028c8620e2b6.png">


### モデルの先頭付近にDepthwise畳み込みがあると学習が遅い
背景: 1x1 convはパラメータ数が少ないもののGPU/TPUをフルに利用できないため遅いらしい。

そのため、efficientnet の基本モジュールをFused-MBConv (3x3 convが入ってる)にかえる。  
Fused-MBConvはパラメータ数が少し増えるものの、推論速度/精度が向上する。

<img width="400" alt="スクリーンショット 2021-05-15 21 37 55" src="https://user-images.githubusercontent.com/19774756/118361357-12e3c080-b5c6-11eb-94bf-443ccdb86a42.png">

<img width="400" alt="スクリーンショット 2021-05-15 21 40 42" src="https://user-images.githubusercontent.com/19774756/118361406-4c1c3080-b5c6-11eb-9037-1fbb3370e831.png">

### モデルを均等にスケールアップするのは最適ではない
どうも、全てのコンポーネントを均等にスケールアップするのは最適ではないらしい。

v2での手法は以下
>V2も基本的にV1のcompound scaling methodに従いながらスケールアップするのですが、次の2つの変更点があります。
テスト時の画像サイズは最大で480
層の数は主に後半のステージで増やす


## 結果
精度は言わずもがなだだけど、parameter数/FLOPSの割にすごい。

<img width="600" alt="スクリーンショット 2021-05-15 21 48 24" src="https://user-images.githubusercontent.com/19774756/118361645-5e4a9e80-b5c7-11eb-8a2e-eaa32d0f27e4.png">


ちなみに、モデルのサイズを大きくするより、データ量を大きくした方が精度のゲインは高い(当然)。  
なので、でかいデータを現実的な時間で学習しきることのできるv2はViTなどよりも優れてると主張している。

[出典リンク](https://qiita.com/omiita/items/1d96eae2b15e49235110)

**Comment by GENZITSU** [comment]
## ドローン x AI で密猟を監視する

背景: ナマコの密猟を監視する必要があるが、夜にやられるため、監視が難しい。

そこでドローン x AIで監視するとのこと。

実は被害金額が結構大きいので、商機あるかも。
>平成２７年には暴力団幹部が関わる大規模な密漁事件が発覚し、約１億７千万円


[出典リンク](https://www.iza.ne.jp/kiji/life/news/210514/lif21051415370028-n1.html)

**Comment by GENZITSU** [comment]
## イオンリテール、店舗にAIカメラシステムを導入、混雑状況の把握や接客、レイアウトの改善に利用

PJT概要

<img width="600" alt="スクリーンショット 2021-05-15 22 13 24" src="https://user-images.githubusercontent.com/19774756/118362522-cbabfe80-b5ca-11eb-8fa1-85c16961f483.png">

できること
- 店内の人数をカウント→混雑状況の検知
- 接客を必要とする顧客の通知
    - 商品を手に取ったかの情報
    - 売り場での滞在時間
- レジ前のカメラで顧客の年代情報を判定する。 (酒類提供のリスクを減らす)
- 性別の推定
- ここら辺のことはマスクの着用有無に関わらず、可能。
- さらに回遊特性を分析することで店舗レイアウトの改善にも役立てるとのこと

受注元は富士通(AIを活用した映像解析ソフトウェア「Fujitsu Technical Computing Solution GREENAGES Citywide Surveillance」)。  
今後、同ソフトウェアに人流シミュレーション機能を追加し、顧客の行動を予測できるようにする。

[出典リンク](https://it.impress.co.jp/articles/-/21491)

**Comment by GENZITSU** [comment]
## torchの__getitem__でnp.random()を使ってはいけない。

DataLoaderの中でnp.random()を使うと以下のような不都合が起きる。  
全てのプロセスで同じtensor idが出現しているのだ...(multi processを行う意味がなくなる...)
```
tensor([[116, 760, 679],   # 1st batch, returned by process 0
        [754, 897, 764]])
tensor([[116, 760, 679],   # 2nd batch, returned by process 1
        [754, 897, 764]])
tensor([[116, 760, 679],   # 3rd batch, returned by process 2
        [754, 897, 764]])

tensor([[866, 919, 441],   # 5th batch, returned by process 0
        [ 20, 727, 680]])
tensor([[866, 919, 441],   # 6th batch, returned by process 1
        [ 20, 727, 680]])
tensor([[866, 919, 441],   # 7th batch, returned by process 2
        [ 20, 727, 680]])
```

## 原因
これはDataloaderがfolkする際に、親プロセスの全ての状態、つまりNumpyの乱数生成器の状態も継承してしまうからである。

## 対処法
`worker_init_fn` を用いて、folkする際にnumpyのシードをずらす。

```
def worker_init_fn(worker_id, epoch):
    np.random.seed(np.random.get_state()[1][0] + worker_id+epoch)

dataset = RandomDataset()
dataloader = DataLoader(dataset, batch_size=2, num_workers=4, 
                        worker_init_fn=worker_init_fn)
```

ちなみに、worker_init_fnにepochを渡さないと、全てのエポック同じ順番でデータを渡すことになる笑

```
epoch: 1
tensor([[282,   4, 785],
        [ 35, 581, 521]])
tensor([[684,  17,  95],
        [774, 794, 420]])
tensor([[939, 988,  37],
        [983, 933, 821]])
-------------------------
epoch: 2
tensor([[282,   4, 785],
        [ 35, 581, 521]])
tensor([[684,  17,  95],
        [774, 794, 420]])
tensor([[939, 988,  37],
        [983, 933, 821]])
-------------------------
```

# 根本的な解決策は`torch.randint`を使うことだそうです
(folkらへんの不具合を考えてくれてるらしい)

[出典リンク](https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects/)

**Comment by GENZITSU** [comment]
## Layout Parserライブラリ

数行でlayout解析ができるようなライブラリが登場した。
<img width="600" alt="スクリーンショット 2021-05-17 20 21 27" src="https://user-images.githubusercontent.com/19774756/118480635-79471b00-b74d-11eb-8a79-fad80e6f00bd.png">

試してみた感じ、日本語ドキュメントには弱そう。英語はそこそこ

<img width="400" alt="スクリーンショット 2021-05-17 20 21 27" src="https://user-images.githubusercontent.com/19774756/118481080-0f7b4100-b74e-11eb-80b0-4462dce9a94e.png">

<img width="400" alt="スクリーンショット 2021-05-17 20 21 27" src="https://user-images.githubusercontent.com/19774756/118481088-11450480-b74e-11eb-8ab9-64a584b29eeb.png">

<img width="400" alt="スクリーンショット 2021-05-17 20 21 27" src="https://user-images.githubusercontent.com/19774756/118481090-12763180-b74e-11eb-8c02-f032512f97e6.png">

<img width="400" alt="スクリーンショット 2021-05-17 20 21 27" src="https://user-images.githubusercontent.com/19774756/118481096-143ff500-b74e-11eb-96b6-f094f3a592ba.png">

[出典リンク](https://layout-parser.github.io/)

**Comment by GENZITSU** [comment]
> ## efficientnet-v2のコードリリース

なぜdepth-wise畳み込みがあまり早くならないかの理由のひとつかもしれない。

>Depthwiseだと演算量が普通のconvに対して1/(チャネル数)となり演算量が減るのですが、activationが巨大でその入出力部分でメモリ律速(GDDR帯域に間に合わない)となってしまいます。そのため最近のGPUはconv2dをDW convに変えても早くならないと思ってます。

https://twitter.com/arutema47/status/1399340659870367745?s=20
